File: src/main/java/no/nav/syfo/service/rest/client/AzureAdTokenConsumer.java
package no.nav.syfo.service.rest.client;

import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.*;
import org.springframework.stereotype.Component;
import org.springframework.util.LinkedMultiValueMap;
import org.springframework.util.MultiValueMap;
import org.springframework.web.client.RestTemplate;
import org.springframework.web.util.UriComponentsBuilder;

import static java.util.Objects.requireNonNull;
import static org.springframework.http.HttpMethod.POST;
import static org.springframework.http.HttpStatus.OK;

@Slf4j
@Component
public class AzureAdTokenConsumer {
    private RestTemplate restTemplateMedProxy;
    private String url;
    private String clientId;
    private String clientSecret;

    public AzureAdTokenConsumer(RestTemplate restTemplateMedProxy,
                                @Value("${aadaccesstoken.url}") String url,
                                @Value("${aad_syfomoteadmin_clientid.username}") String clientId,
                                @Value("${aad_syfomoteadmin_clientid.password}") String clientSecret) {
        this.restTemplateMedProxy = restTemplateMedProxy;
        this.url = url;
        this.clientId = clientId;
        this.clientSecret = clientSecret;
    }

    String getAccessToken(String resource) {
        final HttpHeaders headers = new HttpHeaders();
        headers.setContentType(MediaType.APPLICATION_FORM_URLENCODED);
        final MultiValueMap<String, String> body = new LinkedMultiValueMap<>();
        body.add("client_id", clientId);
        body.add("resource", resource);
        body.add("grant_type", "client_credentials");
        body.add("client_secret", clientSecret);
        final String uriString = UriComponentsBuilder.fromHttpUrl(url).toUriString();
        final ResponseEntity<AzureAdToken> result = restTemplateMedProxy.exchange(uriString, POST, new HttpEntity<>(body, headers), AzureAdToken.class);
        if (result.getStatusCode() != OK) {
            throw new RuntimeException("Henting av token fra Azure AD feiler med HTTP-" + result.getStatusCode());
        }
        return requireNonNull(result.getBody()).getAccess_token();
    }
}

File: src/main/kotlin/no/nav/syfo/azuread/AzureAdTokenConsumer.kt
package no.nav.syfo.azuread

import no.nav.syfo.metric.Metric
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.beans.factory.annotation.Qualifier
import org.springframework.beans.factory.annotation.Value
import org.springframework.http.HttpEntity
import org.springframework.http.HttpHeaders
import org.springframework.http.HttpMethod
import org.springframework.http.MediaType
import org.springframework.stereotype.Component
import org.springframework.util.LinkedMultiValueMap
import org.springframework.util.MultiValueMap
import org.springframework.web.client.RestClientResponseException
import org.springframework.web.client.RestTemplate
import org.springframework.web.util.UriComponentsBuilder
import java.time.Instant
import java.util.*

@Component
class AzureAdTokenConsumer @Autowired constructor(
    private val metric: Metric,
    @param:Qualifier("restTemplateWithProxy") private val restTemplateWithProxy: RestTemplate,
    @param:Value("\${ad.accesstoken.url}") private val url: String,
    @param:Value("\${client.id}") private val clientId: String,
    @param:Value("\${client.secret}") private val clientSecret: String
) {
    private val azureAdTokenMap: MutableMap<String, AzureAdResponse> = HashMap()

    fun accessToken(resource: String): String {
        val omToMinutter = Instant.now().plusSeconds(120L)
        val azureAdResponse = azureAdTokenMap[resource]
        if (azureAdResponse == null || azureAdResponse.expires_on.isBefore(omToMinutter)) {
            try {
                val uriString = UriComponentsBuilder.fromHttpUrl(url).toUriString()
                val result = restTemplateWithProxy.exchange(
                        uriString,
                        HttpMethod.POST,
                        entity(resource),
                        AzureAdResponse::class.java
                )
                azureAdTokenMap[resource] = result.body!!
                metric.countEvent(CALL_AZUREAD_TOKEN_SYSTEM_SUCCESS)
            } catch (e: RestClientResponseException) {
                LOG.error("Request to get token for Azure AD failed with status " + e.rawStatusCode + " and message: " + e.responseBodyAsString)
                metric.countEvent(CALL_AZUREAD_TOKEN_SYSTEM_FAIL)
                throw e
            }
        }
        return azureAdTokenMap[resource]!!.access_token
    }

    fun entity(resource: String): HttpEntity<MultiValueMap<String, String>> {
        val body: MultiValueMap<String, String> = LinkedMultiValueMap()
        body.add("client_id", clientId)
        body.add("resource", resource)
        body.add("grant_type", "client_credentials")
        body.add("client_secret", clientSecret)

        val headers = HttpHeaders()
        headers.contentType = MediaType.APPLICATION_FORM_URLENCODED

        return HttpEntity(body, headers)
    }

    companion object {
        private val LOG = LoggerFactory.getLogger(AzureAdTokenConsumer::class.java)

        private const val CALL_AZUREAD_TOKEN_SYSTEM_BASE = "call_azuread_token_system"
        private const val CALL_AZUREAD_TOKEN_SYSTEM_FAIL = "${CALL_AZUREAD_TOKEN_SYSTEM_BASE}_fail"
        private const val CALL_AZUREAD_TOKEN_SYSTEM_SUCCESS = "${CALL_AZUREAD_TOKEN_SYSTEM_BASE}_success"
    }
}
